mlir/lib/Dialect/Bufferization/Transforms/BufferDeallocationSimplification.cpp

```diff
@@ -314,11 +308,13 @@ struct SplitDeallocWhenNotAliasingAnyOther
 
     SmallVector<Value> remainingMemrefs, remainingConditions;
     SmallVector<SmallVector<Value>> updatedConditions;
-    for (auto [memref, cond] :
-         llvm::zip(deallocOp.getMemrefs(), deallocOp.getConditions())) {
+    for (int64_t i = 0, e = deallocOp.getMemrefs().size(); i < e; ++i) {
+      Value memref = deallocOp.getMemrefs()[i];
+      Value cond = deallocOp.getConditions()[i];
+      SmallVector<Value> otherMemrefs(deallocOp.getMemrefs());
+      otherMemrefs.erase(otherMemrefs.begin() + i);
```

Comment:
This probably even fixes a bug when the same memref is contained in `deallocOp.getMemrefs()` multiple times. The old implementation would allow splitting which would lead to a double-free if I understand correctly without actually testing it or looking into it in detail. Do we have a test-case for that?

Reply:
I tried to write a test for that. We have another pattern that removes duplicate memrefs (`DeallocRemoveDuplicateDeallocMemrefs`). I think whether the bug is triggered or not depends  on the order in which the patterns are applied. I was not able to trigger the double dealloc in practice. (I'll add the test case back anyway, just to be sure.)

Reply:
Makes sense. It's a bit annoying that MLIR doesn't have the utility to easily test individual patterns :/